Simulationsanpassung

jwustmann opened the pull request:
Lösungsvorschlag #66 closes #66

jwustmann committed:
Demand für alle Stunden richtig berechnen

jwustmann committed:
Verzögerung und Demand verbessert

jwustmann committed:
Staubedingte Kapazität jetzt in der richtigen Stunde

nipafx commented:
Du hast dich jetzt an @RF-Buerky's Code orientiert. Das ist keine gute Idee, denn der zäumt das Pferd von hinten auf. Momentan versuchst du, zu Beginn die gesamte Kapazität zu planen. Würdest du wirklich so ein Modell erzeugen? Ist es nicht vernünftiger einen Zeitschritt nach dem nächsten zu simulieren und jeweils zu schauen wo man steht? Siehe auch meinen Kommentar [hier](https://github.com/RF-Buerky/BBFW-Project/issues/64#issuecomment-320953498).

jwustmann commented:
Dann versuche ich es mal von vorne...

nipafx commented:
Guck die `Network` an, [wie es vor der letzten Änderung aussah](https://github.com/RF-Buerky/BBFW-Project/blob/01a2b58ce02885a61d385c3e26d94f281b6ab798/src/main/kotlin/traffic_simulation/RoadNetwork.kt) und stell dir vor, jede Methode hätte einen Parameter `hour` - wie würde das dann aussehen?

RF-Buerky commented:
Werde dich immer im Herzen behalten, mein Code ✝️  💓 #RIP

jwustmann committed:
Simulation jetzt Stunde für Stunde, ohne Map

jwustmann commented:
Das ganze ist jetzt funktionstüchtig, allerdings, solange der Zufallsgenerator (Der mit diesem neuen Design der Berechnung von Demand) noch nicht implementiert ist, löst sich der Stau nicht mehr auf, sobald er mal da ist.

nipafx commented on src/main/kotlin/traffic_simulation/RoadNetwork.kt:
```diff
@@ -2,54 +2,25 @@ package traffic_simulation
+    fun calculateDemandForTimestep(timestep :Int, vehicleList: List<Vehicle>): Int{
+        var demandAtTimestep :Int =0
+        for(vehicle in vehicleList){
+            if(vehicle.wannaDriveInHours.contains(timestep)){
```
Gib dem Fahrzeug eine Methode `wannaDrive(hour: Int)` und ruf die auf. Lass das Fahrzeug darüber entscheiden, wie es sich merkt ob es fahren möchte oder nicht.

jwustmann replied:
ganz ehrlich, this.Vorschlag erschließt sich mir garnicht :,(

nipafx commented on src/main/kotlin/traffic_simulation/RoadNetwork.kt:
```diff
@@ -2,54 +2,25 @@ package traffic_simulation
-                    val newDemand : Int = value + 1
-                    capacityDemand.replace(key, newDemand)
-                }
+    fun calculateDemandForTimestep(timestep :Int, vehicleList: List<Vehicle>): Int{
```
Da hier jetzt nicht nur Verspätung berechnet, sondern auch Stau verursacht werden, sollte der Methodenname angepasst werden.

jwustmann committed:
tests auskommentiert

nipafx commented on src/main/kotlin/traffic_simulation/Vehicle.kt:
```diff
@@ -1,11 +1,111 @@
     fun gettingDelayedInSpecialHour(hour : Int) {
         this.delayedInHours.add(hour)
         delayedInHours.sort()
+        expandDrivingDesireDueToTrafficJam(hour)
```
Aktuell versuchst du den ursprünglichen Fahrbedarf und die Verzögerung im gleichen "Datenbestand" zu verwalten. Versuch die mal, beides getrennt zu merken.

jwustmann replied:
also statt die ursprüngliche Liste zu bearbeiten eine neue erstellen, die die Elemente der ursprünglichen enthält und diese modifizieren?

jwustmann committed:
Lokale Liste funktioniert wieder

nipafx commented:
Wenn du kurz du Arbeit einstellst, schiebe ich einen Commit auf diesen Branch und machen nen Vorschlag.

RF-Buerky commented:
Hast nen ganz schönen Block gemacht für das Anfügen neuer Stunden bei Stau. Vielleicht gehts auch einfacher..

Dazu ein Vorschlag wie man es vielleicht lösen könnte:
Die Klasse Vehicle bekommt das Attribut: `var delayedHours : Int = 0`
(würds nicht in den Konstruktor machen sondern nur in die Klasse)

Immer wenn Stau auftritt wird: `vehicle.delayedHours = vehicle.delayedHours + 1`

Beim durchlaufen durch alle Fahrzeuge prüft man dann, welches Fahrzeug `delayedHours >= 0` hat und gleichzeitig derzeit eigentlich nicht fahren wollte. Wenn dies der Fall ist kann die aktuelle Stunde in die Liste der zu fahrenden Stunden aufgenommen werden und `delayedHours` wird um 1 verringert. 

Bei der nächsten Stunde wird dann wieder eine durch Stau übrig gebliebene Stunde angefügt, solange bis in zukünftigen Stunden irgendwann `delayedHours==0` ist.

jwustmann committed:
ich wüsste nicht, was ich jetzt noch ändern könnte ohne alles zu zerlegen... es funktioniert

jwustmann commented:
@RF-Buerky habe das mit einem riesigen if Block abgefangen, deine counter Variante ginge auch. Aber ich bin heute jetzt nicht mehr vorrätig und überlasse diesen PR euren fähigen Händen

RF-Buerky commented:
Ok, ich werd ma probieren diesen counter einzubinden.

Commit:
Simplify delay computation

nipafx commented:
Too late...

RF-Buerky commented:
Dat is ja fast so wies das Vorhaben war. Das war die Chance Code zu schreiben der gefällt ^^

RF-Buerky commented on src/main/kotlin/traffic_simulation/Vehicle.kt:
```diff
@@ -1,118 +1,29 @@
 package traffic_simulation
 
 class Vehicle(val id: Int, val wannaDriveInHours: MutableList<Int>) {
-    val delayedInHours: MutableList<Int> = mutableListOf() //no vehicle is delayedInHours by default
```
Diese Liste könnte man doch trotzdem beibehalten um am Ende als Output die Infos zu haben inn welchen Stunden ein Vehicle verspätet war.

nipafx replied:
Wenn das Teil des Issues ist (habe ich nicht nachgeguckt), könnte man das machen, ja. :wink:

RF-Buerky replied:
Es gab schon in Block 2 den Auftrag, auszugeben ob eine Verspätung eintraf. 

Und es hieß mal [hier](https://github.com/RF-Buerky/BBFW-Project/pull/41/files/59312840db62967e5e482614517d8b4e363e8e85#diff-1b5c030d3c79b51357f662e7041de7ed)
> I didn't mention replacing it and future tasks actually extend it.

😅

RF-Buerky replied:
Ich versuch mal den Kram wieder einzufügen und hier auf den PullRequest zu schieben.

RF-Buerky replied:
Wie kann man denn auf diesen PullRequest hier pushen? Versuche mit `git push jWusti entire_day` auf den Branch zu pushen von welchem hier der PullRequest ist. `jWusti` ist dabei als Remote der Fork von diesem Projekt. Es gibt aber einen Fehler: `Permission denied`.
@nicolaiparlog

nipafx commented:
Poste mal den Output von `git remote -v`.

RF-Buerky commented:
>$ git remote -v
>jWusti  git@github.com:jwustmann/BBFW-Project.git (fetch)
>jWusti  git@github.com:jwustmann/BBFW-Project.git (push)
>origin  https://github.com/RF-Buerky/BBFW-Project.git (fetch)
>origin  https://github.com/RF-Buerky/BBFW-Project.git (push)


>$ git push jWusti simul
>Permission denied (publickey).
>fatal: Could not read from remote repository.
>
>Please make sure you have the correct access rights
>and the repository exists.

nipafx commented:
Dein Repository ist per HTTPS angebunden (musst du dann nicht jedes Mal Benutzername und Passwort eingeben? 😱 ), Jans per SSH. Wenn du keinen [SSH-Schlüssel hinterlegt hast](https://help.github.com/articles/connecting-to-github-with-ssh/), geht das nicht.

RF-Buerky commented:
Git schient da zu merken wenn ich im Browser in GitHub bin. In dem Fall muss ich mich dann nicht anmelden.

Hast du die Änderungen über den online-Editor gemacht?

RF-Buerky committed:
Liste mit verspäteten Stunden wieder eingefügt

Die entfernte Liste mit allen verspäteten Stunden (`delayedInHours`) eines `Vehicle` wurde wieder eingefügt. Die Liste wird für den Output der verspäteten Stunden gebraucht.

nipafx commented:
> Git schient da zu merken wenn ich im Browser in GitHub bin.

Ah, nice.

> Hast du die Änderungen über den online-Editor gemacht?

Ne, ich hab nen SSH-Schlüssel hinterlegt.

Um das klarzustellen, es liegt nicht an Jans Fork. Ohne SSH-Schlüssel kannst du auch auf deinen eigenen Fork nicht pber `git@...` zugreifen.

Du kannst [die URL umstellen](https://help.github.com/articles/changing-a-remote-s-url/) - die richtige ist https://github.com/jwustmann/BBFW-Project.git

nipafx commented on src/main/kotlin/traffic_simulation/Vehicle.kt:
```diff
@@ -15,6 +16,7 @@ class Vehicle(val id: Int, val wannaDriveInHours: MutableList<Int>) {
         // (otherwise it just waited an hour, not reducing its delay)
         if (wannaDriveInHours.contains(timestep)) {
             delay += 1
+            this.delayedInHours.add(timestep)
```
Wenn das ein echtes Modell wäre, müssten wir jetzt die philosophische Frage beantworten unter welchen Umständen genau das Auto verspätet ist. Machen wir aber nicht, ist also gut so. :smiley:

RF-Buerky replied:
Man ist zu spät wenn man nicht 5min zu früh ist ^^

nipafx commented:
Wird Zeit, die Tests wieder zum Laufen zu bekommen und Travis zufrieden zu stellen. 😛

RF-Buerky commented:
>$ git remote -v
>jWusti  https://github.com/jwustmann/BBFW-Project.git (fetch)
>jWusti  https://github.com/jwustmann/BBFW-Project.git (push)
>origin  https://github.com/RF-Buerky/BBFW-Project.git (fetch)
>origin  https://github.com/RF-Buerky/BBFW-Project.git (push)

>$ git push jWusti entire_day
>error: src refspec entire_day does not match any.
>error: failed to push some refs to 'https://github.com/jwustmann/BBFW-Project.git'

Die Änderung oben hab ich über den online-Editor gemacht..

nipafx commented:
Was sagt `git branch -vv`? (Ja, zwei `v`.)

RF-Buerky commented on src/main/kotlin/traffic_simulation/Vehicle.kt:
```diff
@@ -15,6 +16,7 @@ class Vehicle(val id: Int, val wannaDriveInHours: MutableList<Int>) {
         // (otherwise it just waited an hour, not reducing its delay)
         if (wannaDriveInHours.contains(timestep)) {
             delay += 1
+            this.delayedInHours.add(timestep)
```
Au ja das hab ich an die falsche Stelle gesetzt.

RF-Buerky commented:
>$ git branch -vv
>  master eae3211 [origin/master] Merge pull request #78 from RF-Buerky/csv-weite                      r
>\* simul  7326895 Reformat Code in Vehicle ausgeführt

RF-Buerky committed:
Sortierung von Liste delayedInHours wieder eingefügt

Die Sortierung von der Liste der verspäteten Stunden ist auch wieder eingefügt.

RF-Buerky commented:
Von mir aus kann man das mergen und die Tests von RoadNetworkTest in einer anderen Issue wieder zum laufen bringen.
